crates/cli/src/ts_parser.rs

```diff
@@ -0,0 +1,312 @@
+extern crate swc_common;
+extern crate swc_ecma_parser;
+use anyhow::{bail, Context, Result};
+use std::path::PathBuf;
+
+use swc_common::sync::Lrc;
+use swc_common::SourceMap;
+use swc_ecma_ast::{Decl, Module, ModuleDecl, Stmt, TsInterfaceDecl, TsModuleDecl};
+use swc_ecma_ast::{ModuleItem, TsTypeElement};
+use swc_ecma_parser::{lexer::Lexer, Parser, StringInput, Syntax};
+
+#[derive(Debug, Clone)]
+pub struct Param {
+    pub name: String,
+    pub ptype: String,
+}
+
+impl Param {
+    pub fn new(name: &str, ptype: &str) -> Param {
+        Param {
+            name: name.to_string(),
+            ptype: ptype.to_string().to_uppercase(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Signature {
+    pub name: String,
+    pub params: Vec<Param>,
+    pub results: Vec<Param>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Interface {
+    pub name: String,
+    pub functions: Vec<Signature>,
+}
+
+#[derive(Debug, Clone)]
+pub struct PluginInterface {
+    pub exports: Interface,
+    pub imports: Interface,
+}
+
+/// Parses the "user" part of the module which maps to the wasm imports
+fn parse_user_interface(i: &Box<TsInterfaceDecl>) -> Result<Option<Interface>> {
+    let mut signatures = Vec::new();
+    let name = i.id.sym.as_str();
+    match name {
+        "user" => {
+            for sig in &i.body.body {
+                match sig {
+                    TsTypeElement::TsMethodSignature(t) => {
+                        let name = t.key.as_ident().unwrap().sym.to_string();
+                        let params = t
+                            .params
+                            .iter()
+                            .map(|p| {
+                                let vn = p.as_ident().unwrap().id.sym.as_str();
+                                let typ = p.as_ident().unwrap().type_ann.clone();
+                                let typ = typ.unwrap();
+                                let typ = &typ
+                                    .type_ann
+                                    .as_ts_type_ref()
+                                    .unwrap()
+                                    .type_name
+                                    .as_ident()
+                                    .unwrap()
+                                    .sym;
+                                Param::new(vn, typ)
```

Comment:
I'm sure there is a nicer way to pull out all these values with some high level rust magic but this was what came naturally to me.